### PR TITLE
Desugar parenthesized generic arguments in HIR

### PR DESCRIPTION
**Pull request by petrochenkov**
Fixes ICE in https://github.com/rust-lang/rust/issues/43431 and maybe some other similar issues.

r? @eddyb

### CODE REVIEW

File: src/test/compile-fail/unboxed-closure-sugar-used-on-struct.rs [hunk 1]
```diff
@@ -16,6 +16,7 @@ struct Bar<A> {
 
 fn foo(b: Box<Bar()>) {
     //~^ ERROR parenthesized parameters may only be used with a trait
+    //~| ERROR the type placeholder `_` is not allowed within types on item signatures
```
Comment (petrochenkov) [inline src/test/compile-fail/unboxed-closure-sugar-used-on-struct.rs]: This is an annoying little regression, but from what I tried "infer everything" was the least evil (in terms of redundant diagnostics), given the constraints 1) `usize()` should be a warning and not an error and 2) lifetime resolution has some issues causing ICEs if `()` in non-traits is kept "parenthesized", I hope to fix this in another PR

**Comment by eddyb** [comment]
cc @nikomatsakis I'll take a look later from what I remember is that there were some issues with doing it this way, so we didn't attempt. If only I wrote down the issues...

**Comment by shepmaster** [comment]
@nikomatsakis is on vacation until about 2017-08-15, @eddyb.

**Comment by carols10cents** [comment]
ping @eddyb and/or @nikomatsakis !

File: src/librustc/hir/mod.rs [hunk 2]
```diff
@@ -296,19 +237,33 @@ pub struct AngleBracketedParameterData {
     /// Bindings (equality constraints) on associated types, if present.
     /// E.g., `Foo<A=Bar>`.
     pub bindings: HirVec<TypeBinding>,
+    /// Were parameters written in parenthesized form `Fn(T) -> U`?
+    /// This is required mostly for pretty-printing and diagnostics,
+    /// but also for changing lifetime elision rules to be "function-like".
+    pub parenthesized: bool,
```
Comment (eddyb) [inline src/librustc/hir/mod.rs]: Ah, I see, with this here it's *probably* okay.

**Comment by eddyb** [comment]
@carols10cents I still want to defer to @nikomatsakis here.

**Comment by petrochenkov** [comment]
Rebased.
r? @nikomatsakis

**Comment by nikomatsakis** [comment]
@petrochenkov sorry for delay!

File: src/librustc_driver/driver.rs [hunk 3]
```diff
@@ -833,10 +833,6 @@ pub fn phase_2_configure_and_expand<F>(sess: &Session,
         })
     })?;
 
-    time(time_passes,
-         "early lint checks",
-         || lint::check_ast_crate(sess, &krate));
-
```
Comment (nikomatsakis) [inline src/librustc_driver/driver.rs]: wait what's this -- was this intentionally moved?

**Review by nikomatsakis — COMMENTED** [review]
Sorry, I am going to have to revisit this PR tomorrow. I don't quite understand why it's ok to remove some of the checks that were removed (presumably they are either impossible or taking place earlier, but I have to take another stab at reading it to be sure).

File: src/librustc_driver/driver.rs (hunk 3 above)
Comment (petrochenkov) [inline src/librustc_driver/driver.rs]: Yes, after https://github.com/rust-lang/rust/pull/43522 you can't report late lints without `tcx`, and you can't report early lints after `lint::check_ast_crate` is run. So you can't report lints during AST -> HIR lowering at all.
Thus, I had to move `lint::check_ast_crate` slightly further in the pipeline, to after lowering.

Comment (eddyb) [inline src/librustc_driver/driver.rs]: Makes sense, I'd move all the AST validation into HIR lowering itself.

File: src/librustc_typeck/astconv.rs [hunk 4]
```diff
@@ -951,47 +868,28 @@ impl<'o, 'gcx: 'tcx, 'tcx> AstConv<'gcx, 'tcx>+'o {
...
             }
-            for binding in segment.parameters.bindings() {
+            for binding in &segment.parameters.bindings {
                 self.prohibit_projection(binding.span);
                 break;
             }
         }
     }
 
-    pub fn prohibit_parenthesized_params(&self, segment: &hir::PathSegment, emit_error: bool) {
```
Comment (nikomatsakis) [inline src/librustc_typeck/astconv.rs]: these errors are now reported during lowering? do we have tests for all (or most?) of those cases?

File: src/librustc_typeck/diagnostics.rs [hunk 5]
```diff
@@ -2359,21 +2320,6 @@ impl Foo {
 "##,
      */
 
-E0214: r##"
```
Comment (nikomatsakis) [inline src/librustc_typeck/diagnostics.rs]: this is also reported during lowering now?

**Review by nikomatsakis — APPROVED** [review]
Sorry for the delay. This time when I read through it all made sense to me. r=me presuming that my questions are answered in the affirmative -- the one thing is that I'd like to be sure there are tests for the various error cases... do you know @petrochenkov ?

File: src/librustc_typeck/astconv.rs (hunk 4 above)
Comment (petrochenkov) [inline src/librustc_typeck/astconv.rs]: Yes, everything is reported during lowering now.
There are tests, @qnighy fixed a number of omissions in these checks in https://github.com/rust-lang/rust/pull/41856 and tested everything.

**Pull request merged**